cmd/repo-updater/repos/other_external_services.go

```diff
@@ -0,0 +1,313 @@
+package repos
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sourcegraph/sourcegraph/pkg/api"
+	"github.com/sourcegraph/sourcegraph/pkg/jsonc"
+	"github.com/sourcegraph/sourcegraph/pkg/repoupdater/protocol"
+	"github.com/sourcegraph/sourcegraph/schema"
+	log15 "gopkg.in/inconshreveable/log15.v2"
+)
+
+// OtherReposSyncer periodically synchronizes the configured repos in "OTHER" external service
+// connections with the stored repos in Sourcegraph.
+type OtherReposSyncer struct {
+	// InternalAPI client used to fetch all external servicess and upsert repos.
+	api InternalAPI
+	// RWMutex synchronizing access to repos below
+	mu sync.RWMutex
+	// Latest synced repos cache used by GetRepoInfoByName.
+	repos map[string]*protocol.RepoInfo
+	// Channel passed in NewOtherReposSyncer where synced repos are sent to after being cached.
+	synced chan<- *protocol.RepoInfo
+}
+
+// InternalAPI captures the internal API methods needed for syncing external services' repos.
+type InternalAPI interface {
+	ExternalServicesList(context.Context, api.ExternalServicesListRequest) ([]*api.ExternalService, error)
+	ReposCreateIfNotExists(context.Context, api.RepoCreateOrUpdateRequest) (*api.Repo, error)
+	ReposUpdateMetadata(ctx context.Context, repo api.RepoName, description string, fork, archived bool) error
+}
+
+// NewOtherReposSyncer returns a new OtherReposSyncer. Synced repos will be sent on the given channel.
+func NewOtherReposSyncer(api InternalAPI, synced chan<- *protocol.RepoInfo) *OtherReposSyncer {
+	return &OtherReposSyncer{
+		api:    api,
+		repos:  map[string]*protocol.RepoInfo{},
+		synced: synced,
+	}
+}
+
+// GetRepoInfoByName returns repo info of the repository with the given name.
+func (s *OtherReposSyncer) GetRepoInfoByName(ctx context.Context, name string) *protocol.RepoInfo {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.repos[name]
+}
+
+// Run periodically synchronizes the configured repos in "OTHER" external service
+// connections with the stored repos in Sourcegraph. Termination is done through the passed context.
+func (s *OtherReposSyncer) Run(ctx context.Context, interval time.Duration) error {
+	ticks := time.NewTimer(interval)
```

Comment:
this explains why it only worked the first time

```suggestion
	ticks := time.NewTicker(interval)
```